### ISSUE TITLE
`pushPayload` on `JSONSerializer`

### ISSUE THREAD

**Issue by kaukas** [emberjs/data#4181]
Hello.

Reading the [Ember Models guides](https://guides.emberjs.com/v2.3.0/models/pushing-records-into-the-store/) I get the impression that `store.pushPayload` should be used whenever I want to manually insert server generated data into the store (regardless of adapters or serializers). It mostly delegates the functionality to the underlying serializer. All of the serializers support this except `JSONSerializer`. Is there a reason why `JSONSerializer` can not `pushPayload`?

Thank you!

**Comment by ktrhn** [comment]
I am currently updating an older ember-app. I also was under the impression that JSONSerializer is the way to go. Unfortunately the guides do not tell how to push data from custom ajax requests to the store with JSONSerializer. What works for me until someone has come up with a better solution:
(address is the name of the model, data is a json array):
````
const modelClass = this.store.modelFor('address');
const serializer = this.store.serializerFor('address');
const json = serializer.normalizeArrayResponse(this.store, modelClass, data);
this.store.push(json);
````

**Comment by fivetanley** [comment]
> I also was under the impression that JSONSerializer is the way to go.

I'm curious as to what gave you this impression. I would like to fix this messaging if possible; we definitely don't recommend the JSONSerializer to most people. Usually JSONAPI, REST, and ActiveModel are what I would recommend people reach for first. Our messaging is off here and I apologize for that.

That being said, I don't see why it shouldn't support pushPayload.

**Comment by workmanw** [comment]
I ended up overriding the pushPayload in my serializer and manually calling it.

```javascript
// app/serializers/application.js
import Ember from 'ember';

export default DS.RESTSerializer.extend({
  // If someone invokes push payload, we want to be sure the data is normalized
  pushPayload(store, payload) {
    payload = this._normalizePayloadData(store, payload);
    return this._super(store, payload);
  }
});
```

**Comment by kaukas** [comment]
> I'm curious as to what gave you this impression.

`JSONSerializer` is a better fit for legacy APIs [according to the guides](https://guides.emberjs.com/v2.4.0/models/customizing-serializers/#toc_jsonserializer). And it really is considering that old APIs probably eat and produce JSON that is close to `JSONSerializer`+`EmbeddedRecordsMixin`.

However, there are a couple of issues with this story. In the current app I am working on we started ~ Ember 1.2 (ED was in beta). We used AMS with embedded records (as painful as that was to get to work). I am now upgrading to the latest stable Ember and the AMS JSON structure seems to have (1) changed quite a bit and (2) become incompatible with `EmbeddedRecordsMixin`. I can't use this combination and that's how I ended up with a "legacy API".

As a separate issue, I would not recommend AMS with Rails to anyone. [ActiveModelSerializers has not had a sane stable version](https://github.com/rails-api/active_model_serializers#release-candidate-please-read) for ages (crazy!).

In short, we started with AMS+`EmbeddedRecordsMixin` and this setup became broken for at least 3 reasons. Tried to switch to `JSONSerializer` but `pushPayload` is not defined. Here's my (model specific) version:

```js
      pushPayload: function (store, payload) {
        var modelClass = store.modelFor('model-name');
        var normalizedPayload = this._normalizeResponse(
          store,
          modelClass,
          payload.model_key,
          null, // id
          null, // request type
          !Em.isArray(payload.model_key) // isSingle
        );
        store.push(normalizedPayload);
      }
```

**Comment by bmac** [comment]
It should be possible to use `pushPayload` with the `JSONSerializer` with the caveat that you must specify the type of the record that is being pushed.

```js
store.pushPayload('post', {...});
```

This is because the payload for `JSONSerializer` does not indicate the type of the record(s) being pushed into the store and that makes it hard for Ember Data to infer which serializer it should use.

**Comment by kaukas** [comment]
@bmac, if you do that you will get an error that `pushPayload` is not defined for `JSONSerializer`. If it works then either you created that method yourself or you have some other serializer underneath.

**Comment by bmac** [comment]
My mistake, you are right I do have `pushPayload` defined on my application serializer.

**Comment by fivetanley** [comment]
Can you open a new issue for

>  (2) become incompatible with EmbeddedRecordsMixin

on https://github.com/ember-data/active-model-adapter ?

**Comment by fivetanley** [comment]
@kaukas I think I have figured out some issues with EmbeddedRecordsMixin on the ActiveModelAdapter on a related issue https://github.com/ember-data/active-model-adapter/issues/76

**Comment by kaukas** [comment]
@fivetanley, thank you for the effort. However, I can not test your changes at this point. I believe you can close this issue since I can not follow up on it. In case I can in the future I will re-open this or another issue.

Thank you!

**Comment by wecc** [comment]
@kaukas I'm closing this, feel free to re-open or create a new issue if needed.
